### PR TITLE
Ftr: dubbogo 3.0  Triple support

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
**What this PR does**:
This PR add dubbo3 protocol and dubbo3 net model support, which have features below:
1. use pb to define idl first, not pojo.
2. Perfect compatibility of grpc: 
unray rpc/ two way stream.
3. thread safe.
4. event driven network model.

**Special notes for your reviewer**:
the sample is going to be add. 

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,21 +5,22 @@ require (
 	github.com/RoaringBitmap/roaring v0.5.5
 	github.com/Workiva/go-datastructures v1.0.52
 	github.com/afex/hystrix-go v0.0.0-20180502004556-fa1af6a1f4f5
-	github.com/alibaba/sentinel-golang v0.6.2
+	github.com/alibaba/sentinel-golang v0.6.1
```
Comment (AlexStocks) [inline go.mod:8]: why u decrease sentinel-golang version?

File: protocol/dubbo3/dubbo3_exporter.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+func NewDubbo3Exporter(key string, invoker protocol.Invoker, exporterMap *sync.Map) *Dubbo3Exporter {
+	return &Dubbo3Exporter{
+		BaseExporter: *protocol.NewBaseExporter(key, invoker, exporterMap),
+	}
+}
+
+// Unexport unexport dubbo3 service exporter.
+func (de *Dubbo3Exporter) Unexport() {
+	serviceId := de.GetInvoker().GetUrl().GetParam(constant.BEAN_NAME_KEY, "")
+	interfaceName := de.GetInvoker().GetUrl().GetParam(constant.INTERFACE_KEY, "")
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_exporter.go:46]: why not assign 'de.GetInvoker().GetUrl()' to a new var to decrease the invocation time?

File: protocol/dubbo3/dubbo3_invoker.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+// Invoke call remoting.
+func (di *Dubbo3Invoker) Invoke(ctx context.Context, invocation protocol.Invocation) protocol.Result {
+	var (
+		result protocol.RPCResult
+	)
+
+	var in []reflect.Value
+	in = append(in, reflect.ValueOf(ctx))
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_invoker.go:83]: using 'in := append(in, reflect.ValueOf(ctx))' instead

File: protocol/dubbo3/dubbo3_invoker.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	} else {
+		_ = hessian2.ReflectResponse(res[0], invocation.Reply())
+	}
+
+	return &result
+}
+
+// get timeout including methodConfig
+func (di *Dubbo3Invoker) getTimeout(invocation *invocation_impl.RPCInvocation) time.Duration {
+	var timeout = di.GetUrl().GetParam(strings.Join([]string{constant.METHOD_KEYS, invocation.MethodName(), constant.TIMEOUT_KEY}, "."), "")
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_invoker.go:105]: delete the var and using `:=` instead.

File: protocol/dubbo3/dubbo3_protocol.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dubbo3
+
+import (
+	"fmt"
+	"google.golang.org/grpc"
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_protocol.go:21]: move it to the second import block.

File: protocol/dubbo3/dubbo3_protocol.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	dubbo3Protocol *Dubbo3Protocol
+)
+
+// It support dubbo protocol. It implements Protocol interface for dubbo protocol.
+type Dubbo3Protocol struct {
+	protocol.BaseProtocol
+	// It is store relationship about serviceKey(group/interface:version) and ExchangeServer
+	// The ExchangeServer is introduced to replace of Server. Because Server is depend on getty directly.
+	serverMap  map[string]*dubbo3.TripleServer
+	serverLock sync.Mutex
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_protocol.go:56]: pls move the lock and place it above its guard object.

File: protocol/dubbo3/dubbo3_protocol.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/apache/dubbo-go/remoting/dubbo3"
+)
+
+const (
+	// DUBBO3 is dubbo3 protocol name
+	DUBBO3 = "dubbo3"
+)
+
+func init() {
+	logger.Warn("Init extension")
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_protocol.go:42]: write log in init?

File: protocol/dubbo3/dubbo3_protocol.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	// GetProxyImpl gets proxy.
+	GetProxyImpl() protocol.Invoker
+	// ServiceDesc gets an RPC service's specification.
+	ServiceDesc() *grpc.ServiceDesc
+}
+
+// openServer open a dubbo3 server
+func (dp *Dubbo3Protocol) openServer(url *common.URL) {
+	_, ok := dp.serverMap[url.Location]
+	if !ok {
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_protocol.go:116]: pls do not write js-style if.

File: protocol/dubbo3/dubbo3_protocol.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+			in = append(in, reflect.ValueOf(invoker))
+			m.Func.Call(in)
+
+			srv := dubbo3.NewTripleServer(url, service)
+
+			dp.serverMap[url.Location] = srv
+
+			srv.Start()
+		}
+		dp.serverLock.Unlock()
```
Comment (AlexStocks) [inline protocol/dubbo3/dubbo3_protocol.go:152]: using defer unlock, pls.

File: protocol/dubbo3/impl/header.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+)
+
+func init() {
+	remoting.SetProtocolHeaderHandler("dubbo3", NewTripleHeaderHandler)
+}
+
+// TripleHeader define the h2 header of triple impl
+type TripleHeader struct {
+	method         string
+	streamID       uint32
```
Comment (AlexStocks) [inline protocol/dubbo3/impl/header.go:41]: lower  case? why?

File: protocol/dubbo3/impl/header.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+			tripleHeader.TracingID = f.Value
+		case "tri-trace-rpcid":
+			tripleHeader.TracingRPCID = f.Value
+		case "tri-trace-proto-bin":
+			tripleHeader.TracingContext = f.Value
+		case "tri-unit-info":
+			tripleHeader.ClusterInfo = f.Value
+		case "content-type":
+			tripleHeader.ContentType = f.Value
+		case "grpc-encoding":
```
Comment (AlexStocks) [inline protocol/dubbo3/impl/header.go:131]: remark them, pls.

File: protocol/dubbo3/impl/package.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+// TriplePackageHandler handles package of triple
+// e.g. now it impl as deal with pkg data as: [:5]is length and [5:lenght] is body
+type TriplePackageHandler struct {
+	header *TripleHeader
+	//codec  *CodeC
+}
+
+func (t *TriplePackageHandler) Frame2PkgData(frameData []byte) []byte {
+	lineHeader := frameData[:5]
```
Comment (AlexStocks) [inline protocol/dubbo3/impl/package.go:36]: pls check the length of @frameData firstly.

File: protocol/dubbo3/impl/package.go [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+}
+func (t *TriplePackageHandler) Pkg2FrameData(pkgData []byte) []byte {
+	rsp := make([]byte, 5+len(pkgData))
+	rsp[0] = byte(0)
+	binary.BigEndian.PutUint32(rsp[1:], uint32(len(pkgData)))
+	copy(rsp[5:], pkgData[:])
+	return rsp
+}
+
+func init() {
```
Comment (AlexStocks) [inline protocol/dubbo3/impl/package.go:48]: pls move the init to the header of this file.

File: remoting/dubbo3/constant.go [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dubbo3
+
+const defaultRWBufferMaxLen = 4096
```
Comment (AlexStocks) [inline remoting/dubbo3/constant.go:20]: firstly, u should place the tow constants into one const () block.

sencondly, pls set the constant type of defaultRWBufferMaxLen.

File: remoting/dubbo3/dubbo3_client.go [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+func newTripleConn(client *TripleClient) *TripleConn {
+	return &TripleConn{
+		client: client,
+	}
+}
+
+// getInvoker return invoker that have service method
+func getInvoker(impl interface{}, conn *TripleConn) interface{} {
+	var in []reflect.Value
+	in = append(in, reflect.ValueOf(conn))
```
Comment (AlexStocks) [inline remoting/dubbo3/dubbo3_client.go:73]: using :=

File: remoting/dubbo3/sstream.go [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dubbo3
+
+import (
```
Comment (AlexStocks) [inline remoting/dubbo3/sstream.go:20]: split it pls.

File: remoting/dubbo3/stream.go [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dubbo3
+
+import (
```
Comment (AlexStocks) [inline remoting/dubbo3/stream.go:20]: split it, pls.

File: go.mod [hunk 18]
```diff
@@ -5,21 +5,22 @@ require (
...
 	github.com/elazarl/go-bindata-assetfs v1.0.0 // indirect
 	github.com/emicklei/go-restful/v3 v3.0.0
 	github.com/frankban/quicktest v1.4.1 // indirect
 	github.com/fsnotify/fsnotify v1.4.7
 	github.com/go-co-op/gocron v0.1.1
 	github.com/go-resty/resty/v2 v2.3.0
+	github.com/golang/protobuf v1.3.1
 	github.com/golang/mock v1.4.4
-	github.com/golang/protobuf v1.4.3
+	github.com/golang/protobuf v1.4.0
```
Comment (fangyincheng) [inline go.mod:23]: why do you reduce its version

File: protocol/dubbo3/dubbo3_exporter.go (hunk 2 above)
Comment (LaurenceLiZhixin) [reply protocol/dubbo3/dubbo3_exporter.go:46]: I just copied this part from original dubbo. I'll fix it now in dubbo3.

File: protocol/dubbo3/dubbo3_invoker.go (hunk 3 above)
Comment (LaurenceLiZhixin) [reply protocol/dubbo3/dubbo3_invoker.go:83]: r u sure?
I guess change 
`var in []reflect.Value `
to
`in := make(reflect.Value, 0, 16)`
may be better?

Comment (AlexStocks) [reply protocol/dubbo3/dubbo3_invoker.go:83]: change it as ur idae

File: go.mod (hunk 18 above)
Comment (LaurenceLiZhixin) [reply go.mod]: fixed!

**Review by Patrick0308 — COMMENTED** [review]
May be `remoting/dubbo3` can move to a new project?

File: protocol/dubbo3/dubbo3_exporter.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (cityiron) [inline protocol/dubbo3/dubbo3_exporter.go:1]: I think dubbo3_exporter.go should change to dubbo_exporter.go, like other protocol, the package dubbo3 is enough.

**Pull request closed without merging**